### PR TITLE
Allow normal program exit to be detected under MINIMAL_RUNTIME

### PR DESCRIPTION
**Pull request by sbc100**
Followup to #14586 which allowed the test framework to detect `exit()`
calls with MINIMAL_RUNTIME + EXIT_RUNTIME.    This change extends
that support to normal program exit (returning from main)  which was previously
undetectable.

We also add a `window.unhandledrejection` handler to
`browser_reporting.js` since under MINIMAL_RUNTIME that main function
runs in the `then()` handler of `WebAssembly.instantiate` and so any
exceptions thown from there show up as unhandled promise rejections.
This change is generally useful, and enables us to detech more types
of errors during testing.

I've also moved error handling code for MINIMAL_RUNTIME from a
`WebAssembly.instantiate.then().catch()` to the `onerror` handler of the
`WebAssembly.instantiate.then()`.  This means that is will now catch
only errors during `WebAssembly.instantiate` and not excpetions throwns
by user code after instantiation.  This avoids the JS fallback handling
code in the case the application throws and exceptions or calls exit().

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Is this related to uncatchable exceptions in https://github.com/emscripten-core/emscripten/issues/14777 ?

**Comment by sbc100** [comment]
Not really related although they both related to unhandled promise rejections.    

I think the thing you call uncatchable is an unhandled promise and you can register and global unhandlerrejection handler in your app if you want to catch it.

**Comment by vadimkantorov** [comment]
I'm using modularized modules and await module instantiation. It would be better if exceptions occurred were catchable with regular ways without setting up a global handler.

**Comment by sbc100** [comment]
I don't think this change makes it any hard or easier to use try/catch blocks to catch exceptions.

As it happens the test framework prefers to use the global handler, but this change doesn't effect how easy or hard it is to catch using other methods.

**Comment by sbc100** [comment]
PTAL, which change unblocked converting more of our test suite from using REPORT_RESULT.

File: tests/common.py [hunk 1]
```diff
@@ -400,7 +400,7 @@ def setUp(self):
     super().setUp()
     self.settings_mods = {}
     self.emcc_args = ['-Werror']
-    self.node_args = ['--stack-trace-limit=50', '--unhandled-rejections=throw']
+    self.node_args = ['--stack-trace-limit=50', '--trace-uncaught', '--unhandled-rejections=throw']
```
Comment (kripken) [inline tests/common.py:403]: unintended?

File: tests/test_core.py [hunk 2]
```diff
@@ -1030,19 +1030,15 @@ def test_exceptions_minimal_runtime(self):
     self.set_setting('EXIT_RUNTIME')
     self.maybe_closure()
     self.set_setting('MINIMAL_RUNTIME')
+    self.emcc_args += ['--pre-js', test_file('minimal_runtime_support.js')]
```
Comment (kripken) [inline tests/test_core.py:1033]: why is this needed here but not in other minimal runtime tests?

**Review by kripken — COMMENTED** [review]
Looks like this should have no code size impact by default? (only when opting in to EXIT_RUNTIME)

File: tests/common.py (hunk 1 above)
Comment (sbc100) [reply tests/common.py:403]: Its somewhat unrelated but also makes node error message more useful in general.  See https://nodejs.org/api/cli.html#cli_trace_uncaught.  Basically without this node prints stack traces for uncaught Error object but not the full stack trace for other kinds of objects.   I could put this in a separate PR if you prefer but I think its OK here.

**Comment by sbc100** [comment]
> Looks like this should have no code size impact by default? (only when opting in to EXIT_RUNTIME)

Yes, exactly.

File: tests/test_core.py (hunk 2 above)
Comment (sbc100) [reply tests/test_core.py:1033]: I think it would be good to add both `EXIT_RUNTIME` and this support file to all `MINIMAL_RUNTIME` tests, but its not strictly needed today.

If you look at `MINIMAL_RUNTIME` today, without this change, calls to `exit()` (in default builds) from inside of main are silently logged and node returns 0 (even for non-zero exit() calls).   Also, return codes from main are silently ignored. 

With this change calls to exit() are no longer ignored (due to the then.catch() change) and returns from main are still ignored (unless one sets `EXIT_RUNTIME`).

So basically any test that needs to accurately detect the exit code should set `EXIT_RUNTIME` and use `minimal_runtime_support.js`... I'm guessing these are the only tests that use `MINIMAL_RUNTIME` and expect a non zero return code.

File: tests/common.py (hunk 1 above)
Comment (kripken) [reply tests/common.py:403]: I guess it's ok here. It does say it may make things slower though, in that link? Is it worth running all the time then?

File: tests/test_core.py (hunk 2 above)
Comment (kripken) [reply tests/test_core.py:1033]: I see. Sounds reasonable. How about renaming `minimal_runtime_support.js`, though, to a name that indicates this is for support of EXIT_RUNTIME specifically?

Comment (sbc100) [reply tests/test_core.py:1033]: Done

File: emcc.py [hunk 3]
```diff
@@ -2031,7 +2031,10 @@ def check_memory_setting(setting):
     if options.shell_path == shared.path_from_root('src', 'shell.html'):
       options.shell_path = shared.path_from_root('src', 'shell_minimal_runtime.html')
 
-    if settings.ASSERTIONS and settings.MINIMAL_RUNTIME:
+    if settings.EXIT_RUNTIME:
+      settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE += ['proc_exit']
+
+    if settings.ASSERTIONS:
       # In ASSERTIONS-builds, functions UTF8ArrayToString() and stringToUTF8Array() (which are not JS library functions), both
       # use warnOnce(), which in MINIMAL_RUNTIME is a JS library function, so explicitly have to mark dependency to warnOnce()
```
Comment (juj) [inline emcc.py:2039]: This PR changes the logic to remove MINIMAL_RUNTIME only check for adding `$warnOnce` (which I suppose is ok), but the comment still reads MINIMAL_RUNTIME specific. Although I suppose technically the comment is still right. Maybe it makes sense to adjust the comment to highlight the fact that `$warnOnce` is now added always for simplicity (I presume that is the rationale here?) even though needed just for MINIMAL_RUNTIME.

This change does have the effect of emitting `function warnOnce()` twice to non-MINIMAL_RUNTIME builds though, once via the preamble and a second time via the JS function. Not sure how well that will DCE via Closure, since it has problems optimizing away stuff from global JS scope.

**Review by juj — APPROVED** [review]
Looks good!

**Comment by juj** [comment]
Or actually on second thought, I wonder if instead of this throw, the test harness could register an `<<< ATEXITS >>>` handler that throws? Not all users might like to have a `throw` even if they are using EXIT_RUNTIME? (it doesn't seem particularly useful for end users?)

File: emcc.py (hunk 3 above)
Comment (sbc100) [reply emcc.py:2039]: Sorry I should have mentioned that this entire block is already wrapped in `if MINIMAL_RUNTIME` (see line 2029) so this was a redundant check.   Its not obvious when reading the diff

**Comment by sbc100** [comment]
> Or actually on second thought, I wonder if instead of this throw, the test harness could register an `<<< ATEXITS >>>` handler that throws? Not all users might like to have a `throw` even if they are using EXIT_RUNTIME? (it doesn't seem particularly useful for end users?)

Yes I did consider that option, but given that `exit(0)` already thows I thought this would more consistent.  Also `exit()` under `MINIMAL_RUNTIME` doesn't call `<<< ATEXITS >>>` today (thats another issue) so we would need to handle both the exit() exceptions and the `<<< ATEXITS >>>` handler in the test framework.  

Another way of putting it: any user of `MINIMAL_RUNTIME` + `EXIT_RUNTIME` already has to handle `exit()` exceptions that get thrown by `exit(0)` so I think this just makes `return 0` and `exit(0)` consistent.

**Review by kripken — APPROVED** [review]
lgtm aside from the question of performance of trace-uncaught, which I think might be worth measuring first.

File: tests/common.py (hunk 1 above)
Comment (sbc100) [reply tests/common.py:406]: I'll revert this here and propose it as a separate PR

Comment (sbc100) [reply tests/common.py]: https://github.com/emscripten-core/emscripten/pull/14890
